packages/react-storage/src/components/StorageBrowser/Views/LocationActionView/useHandleUpload.ts

```diff
@@ -129,60 +152,148 @@ export function useHandleUpload({
     setTasks((prevTasks) =>
-      prevTasks.map(({ key, data, progress, size }) => {
-        const { bucket: bucketName, credentialsProvider, region } = getConfig();
-        const input: UploadDataWithPathInput = {
-          path: `${prefix}${key}`,
-          data,
-          options: {
-            bucket: { bucketName, region },
-            locationCredentialsProvider: credentialsProvider,
-            onProgress: ({ totalBytes, transferredBytes }) => {
-              const progress = totalBytes ? transferredBytes / totalBytes : 0;
-              const nextTask: TaskUpdate = {
-                key,
-                progress,
-                ...(totalBytes === transferredBytes
-                  ? { cancel: undefined }
-                  : undefined),
-              };
-              setTasks((curr) => updateTasks(curr, nextTask));
-            },
-            preventOverwrite,
-          },
-        };
+      prevTasks.map((task) => ({
+        ...task,
+        status: 'QUEUED',
+        cancel: () =>
+          setTasks((prevTasks) =>
+            updateTasks(prevTasks, {
+              key: task.key,
+              status: 'CANCELED',
+              cancel: undefined,
+            })
+          ),
+      }))
+    );
 
-        const { cancel, result } = uploadData(input);
+    const uploadNext = async () => {
+      const currentTasks = tasksRef.current;
 
-        result
-          .then(() => {
-            setTasks((curr) => updateTasks(curr, { key, status: 'COMPLETE' }));
+      if (currentIndex >= currentTasks.length) {
+        return;
+      }
+
+      const queuedTask = currentTasks[currentIndex];
+
+      if (queuedTask.status === 'CANCELED') {
+        currentIndex++;
+
+        uploadNext();
+
+        return;
+      }
+
+      const pendingTask = processUpload(queuedTask);
+
+      currentIndex++;
+
+      const { task, result } = pendingTask;
+      const { key } = task;
+
+      try {
+        await result;
+
+        setTasks((prevTasks) =>
+          updateTasks(prevTasks, { key, status: 'COMPLETE' })
+        );
+      } catch (error) {
+        setTasks((prevTasks) =>
+          updateTasks(prevTasks, {
+            key,
+            status: isCancelError(error) ? 'CANCELED' : 'FAILED',
+            cancel: undefined,
           })
-          .catch(() => {
-            setTasks((curr) => updateTasks(curr, { key, status: 'FAILED' }));
-          });
+        );
+      } finally {
+        uploadNext();
+      }
+    };
 
-        const handleCancel = () => {
-          cancel();
+    // Start the initial set of uploads
+    const initialUploads = Array(batchSize)
+      .fill(null)
+      .map(() => uploadNext());
 
-          setTasks((curr) =>
-            // assign cancel to noop
-            updateTasks(curr, { key, cancel: () => null, status: 'CANCELED' })
-          );
-        };
+    await Promise.all(initialUploads);
+  };
 
+  const handleCancel = () => {
```

Comment:
We should just be able to iterate over the current `tasks` here and call `cancel` for the cancelable tasks without the additional logic i think?

Reply:
that's true, I could simplify the logic in here. I think we just need to check if `cancel` is the cancel function, right?

Reply:
Think yes? Something like `tasks.forEach(({ cancel }) => isFunction(cancel) ? cancel() : undefined)` maybe?

Reply:
We would still need to update the state to show the tasks got canceled too right?